sycl/doc/extensions/InvokeSIMD/SYCL_INTEL_invoke_simd.asciidoc

```diff
@@ -0,0 +1,364 @@
+types (e.g. +nd_item+) being passed between SPMD and SIMD contexts.  This
+restriction prevents a SIMD function from calling functions that are only
+well-defined in SPMD contexts (e.g. sub-group barriers).  If a SIMD function
+requires access to a member variable of a SYCL class, the value of the
+variable should be passed via a dedicated argument (e.g. the value returned
+by +sub_group::get_local_id()[0]+ could be passed as an integer to a
++Callable+ expecting a +sycl::ext::oneapi::simd<uint32_t, N>+).
+
+The +invoke_simd+ function has the same requirements as other group functions
+(as defined in Section 4.17.3 of the SYCL 2020 specification). A call to
++invoke_simd+ must be encountered in converged control flow by all work-items
+in the group, and the call acts as a synchronization point -- the +Callable+ is
+not invoked until all work-items reach the call to +invoke_simd+, and all
+work-items must wait for the +Callable+ to complete before continuing
+execution.
+
+The example below shows how to invoke a simple SIMD function that scales all
+elements of a SIMD type +x+ by a scalar value +n+, both with and without a
+SIMD mask parameter.
+
+[source, c++]
+----
+sycl::ext::oneapi::simd<float, 8> scale(sycl::ext::oneapi::simd<float, 8> x, float n)
+{
+  return x * n;
+}
+
+sycl::ext::oneapi::simd<float, 8> masked_scale(sycl::ext::oneapi::simd<float, 8> x, float n, sycl::ext::oneapi::simd_mask<bool, 8> mask)
+{
+  std::experimental::where(mask, x) *= n;
+  return x;
+}
+
+q.parallel_for(..., sycl::nd_item<1> it) [[sycl::reqd_sub_group_size(8)]]
+{
+  sycl::sub_group sg = it.get_sub_group();
+  float x = ...;
+  float n = ...;
+
+  // invoke SIMD function
+  // x values from each work-item are combined into a simd<float, 8>
+  float y = sycl::ext::oneapi::invoke_simd(sg, scale, x, sycl::ext::oneapi::uniform(n));
+
+  // invoke SIMD function with a mask parameter
+  // x values from each work-item be a simd<float, 8>
+  // mask values from each work-item are combined into a simd_mask<bool, 8>
+  bool mask = (it.get_local_id(0) % 2);
+  float z = sycl::ext::oneapi::invoke_simd(sg, masked_scale, x, sycl::ext::oneapi::uniform(n), mask);
+});
+----
+
+==== Execution Model
+
+Execution of a SIMD function must produce the same result as-if it had been
+executed by a single work-item executing SIMD instructions.
+
+The following are all examples of valid implementations of this model:
+
+- Mapping SIMD operations directly to SIMD instructions.
+
+- Mapping logical SIMD operations wider than the native SIMD width to multiple
+  SIMD instructions.
+
+- Mapping each element of a SIMD type to an individual work-item in a
+  sub-group, with appropriate barriers and fences generated between SIMD
+  operations.
+
+- Mapping SIMD operations to an unrolled loop executed by a single work-item.
+
+Since this execution model guarantees SIMD-like behavior, there is no need for
+the user to insert sub-group barriers in an explicit SIMD function.
```

Comment:
I guess this is also impossible, because you can only obtain a `sub_group` handle through `nd_item`, which can't be easily passed down to `SIMD` context (because it is not an arithmetic type and it is not uniform across sub-group) and that handle is required as the first argument of `group_barrier` free function - right?

Reply:
That's right.  The intent of this paragraph was to make it clear to users that they no longer have to worry about barriers, and to make it clear to implementers that if their implementation requires barriers between statements then the user shouldn't be responsible for generating them.  Would the below be clearer:
> Since this execution model guarantees SIMD-like behavior, there is no need for the user to insert any form of explicit synchronization functions to ensure memory consistency across SIMD lanes.

Reply:
> Would the below be clearer:

The existing wording is fine by me, I just wanted to double-check my understanding and my comment could also be treated as a "prequel"  to [which SYCL APIs can be used within SIMD context?](https://github.com/intel/llvm/pull/3445#discussion_r607864083) question asked below

Reply:
I've changed the wording anyway (in 18f3232b186e).  Answering your question made me realize that mentioning sub-group barriers here is unnecessary, and I'm worried that it could lead to confusion.